### PR TITLE
feat: add Failed status for cloud provider

### PR DESCRIPTION
**Pull request by xiaohan1202**

### CODE REVIEW

File: controllers/cluster/api/v1/cluster_types.go [hunk 1]
```diff
@@ -82,6 +82,7 @@ type ClusterStatus struct {
 
 //+kubebuilder:object:root=true
 //+kubebuilder:subresource:status
+//+kubebuilder:printcolumn:name="Status",type="string",JSONPath=".status.status",description="Status of Cluster in group"
```
Comment (cuisongliu) [inline controllers/cluster/api/v1/cluster_types.go:85]: Phase is better ?

File: controllers/infra/controllers/infra_controller.go [hunk 2]
```diff
@@ -251,6 +269,44 @@ func getSecretName(infra *infrav1.Infra) string {
...
+			return err
+		}
+		return nil
+	}); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (r *InfraReconciler) updateFinalizers(ctx context.Context, objectKey client.ObjectKey, obj client.Object, finalizers []string) error {
```
Comment (cuisongliu) [inline controllers/infra/controllers/infra_controller.go:289]: see user controller usage.

File: controllers/cluster/api/v1/cluster_types.go (hunk 1 above)
Comment (xiaohan1202) [reply controllers/cluster/api/v1/cluster_types.go:85]: this format is the same as infra CRD

File: controllers/infra/controllers/infra_controller.go [hunk 3]
```diff
@@ -91,15 +92,22 @@ func (r *InfraReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl
 		return ctrl.Result{}, err
 	}
 
+	// if infra status is failed, delete finalizer and do not reconcile
+	if infra.Status.Status == v1.Failed.String() {
```
Comment (cuisongliu) [inline controllers/infra/controllers/infra_controller.go:96]:
```
if _, err := r.finalizer.RemoveFinalizer(ctx, infra, r.DeleteInfra); err != nil {
			return ctrl.Result{}, err
		}
		return ctrl.Result{}, nil
```

**Review by cuisongliu — APPROVED** [review]

**Pull request merged**